include/nuttx/sched.h

```diff
@@ -748,6 +748,10 @@ struct tcb_s
 #if CONFIG_TASK_NAME_SIZE > 0
   char name[CONFIG_TASK_NAME_SIZE + 1];  /* Task name (with NUL terminator)     */
 #endif
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+  int syscall_nest;                      /* Syscall nest level */
```

Comment:
Nobody use this variable?

Reply:
sched_note_syscall_enter() and _leave() use it.
It is used to exclude the nested system call in the trace data (ex.  sched_lock/unlock called by another kernel APIs)

Reply:
From the application developer's point of view, what and when system calls are used is important, but the internal of the system call is not.
And, the most important reason is, the trace viewer "tracecompass" cannot handle nested system call well.

Reply:
After search the code base, all arch use the same name(nsyscalls), so we use tcb->xcpt.nsyscalls directly which simplify the code a lot.

Reply:
Unfortunately tcb->xcpt.nsyscalls is not fit for my expectation.
It can be used only in PROTECTED and KERNEL build and it counts the nesting level of user mode -> kernel mode transition.
Because my sched_note_syscall_enter/leave implementation hooks the function call in the kernel space directly, the required count is the nesting level of syscall function calls and it also needs in FLAT build too.
("syscall function" means the kernel APIs listed in syscall/syscall.csv)

Reply:
I am still thinking that it's very useful to trace the nested syscall because it could show the real sequence. Could you explain what's wrong with tracecompass in this case. BTW, we can filter out the nested syscall if tracecompass can't handle it correctly.

Reply:
@xiaoxiang781216 
This is a screenshot of tracecompass to compare the case of removing the nested syscall (upper) and not removing (lower).
The real execution time of putenv() is upper one, but when the nested syscall is not removed, tracecompass misjudges the end of syscall and displays incorrect syscall execution period.

However, I understand your thought that it's very useful to trace the nested syscall.
So I have changed my mind and will propose the alternative idea for it:
- sched_note.c records the syscall notes as it is.
- The app code which translates the note record into ftrace text format counts the nesting level of syscall and removes the nested syscalls for correctly tracecompass display.

Because the note data would be available by not only tracecompass but another tools, it would be better to remove the special treatment for tracecompass from kernel code.
How about this idea?

![tracecompass-fig1](https://user-images.githubusercontent.com/36463524/94888914-890f3900-04b5-11eb-89c3-1e56baa0f2e6.png)

Reply:
Yes, this is what I suggest: we can do the filter in the userspace or even on PC before we feed the trace to tracecompass.